service/tablet_allocator.cc

```diff
@@ -132,6 +132,85 @@ class load_balancer_stats_manager {
     }
 };
 
+// Used to compare different migration choices in regard to impact on load imbalance.
+// There is a total order on migration_badness such that better migrations are ordered before worse ones.
+struct migration_badness {
+    double src_shard_badness = 0;
+    double src_node_badness = 0;
+    double dst_shard_badness = 0;
+    double dst_node_badness = 0;
+    bool bad;
+
+    migration_badness()
+        : bad(false)
+    {}
+
+    migration_badness(double src_shard_badness, double src_node_badness, double dst_shard_badness, double dst_node_badness)
+        : src_shard_badness(src_shard_badness)
+        , src_node_badness(src_node_badness)
+        , dst_shard_badness(dst_shard_badness)
+        , dst_node_badness(dst_node_badness)
+        , bad(src_shard_badness > 0 || src_node_badness > 0 || dst_shard_badness > 0 || dst_node_badness > 0)
+    {}
+
+    double node_badness() const {
+        return std::max(src_node_badness, dst_node_badness);
+    }
+
+    double shard_badness() const {
+        return std::max(src_shard_badness, dst_shard_badness);
+    }
+
+    bool is_bad() const {
+        return bad;
+    }
+
+    bool operator<(const migration_badness& other) const {
+        // Prefer candidates with no across-node badness to those with across-node badness.
+        // Then, prefer those with lowest shard badness.
+        // We want to balance nodes first as balancing nodes internally between shards is cheap.
+        if (node_badness() > 0 || other.node_badness() > 0) {
+            return node_badness() < other.node_badness();
```

Comment:
If their equally bad, we should compare shard_badness. But maybe that breaks total ordering.

Suppose node_badness cannot be zeroed. How will we compare shard badness?

If badness can be quantified, tiny differences should be ignored. If we have small node inequality but large shard inequality, shouldn't we fix shard inequality first?

Perhaps compare abs(delta(node badness)) to 5 abs(delta(shard badness)) to see what to measure?

Reply:
Agree about equally bad node badness, we could compare shard badness in that case. Though it seems like a corner case we could do without.

> If we have small node inequality but large shard inequality, shouldn't we fix shard inequality first?

You mean from the perspective of the global algorithm, or this metric? This metric deals with candidate selection. The idea is that by minimizing node badness, we also minimize final shard badness post-intra-node-rebalancing. Going against node badness would always increase shard badness, because overcommiting a node also overcommits shards. So I don't think we should ignore tiny differences in node badness if our goal is to minimize the final shard badness.

The supporting step so it have intra-node rebalancing which would fix shard badness within the node. We don't have that yet.

> Perhaps compare abs(delta(node badness)) to 5 abs(delta(shard badness)) to see what to measure?

I don't understand what you mean here.